x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,12 +107,87 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
+
+        if (optimizedReadSize < length) {
+            // we did not read everything in an optimized fashion, so read the remainder directly
+            try (InputStream inputStream
+                     = blobContainer.readBlob(fileInfo.partName(part), pos + optimizedReadSize, length - optimizedReadSize)) {
+                final int directReadSize = inputStream.read(b, offset + optimizedReadSize, length - optimizedReadSize);
+                assert optimizedReadSize + directReadSize == length : optimizedReadSize + " and " + directReadSize + " vs " + length;
+                position += directReadSize;
+            }
+        }
+    }
+
+    /**
+     * Attempt to satisfy this read in an optimized fashion using {@code streamForSequentialReadsRef}.
+     * @return the number of bytes read
+     */
+    private int readOptimized(int part, long pos, byte[] b, int offset, int length) throws IOException {
+        if (sequentialReadSize == NO_SEQUENTIAL_READ_OPTIMIZATION) {
+            return 0;
+        }
+
+        int read = 0;
+        if (streamForSequentialReads == null) {
+            // starting a new sequential read
+            read = readFromNewSequentialStream(part, pos, b, offset, length);
+        } else if (streamForSequentialReads.canContinueSequentialRead(part, pos)) {
+            // continuing a sequential read that we started previously
+            read = streamForSequentialReads.read(b, offset, length);
+            if (streamForSequentialReads.isFullyRead()) {
+                // the current stream was exhausted by this read, so it should be closed
+                streamForSequentialReads.close();
+                streamForSequentialReads = null;
+            } else {
+                // the current stream contained enough data for this read and more besides, so we leave it in place
+                assert read == length : length + " remaining";
+            }
+
+            if (read < length) {
+                // the current stream didn't contain enough data for this read, so we must read more
+                read += readFromNewSequentialStream(part, pos + read, b, offset + read, length - read);
+            }
+        } else {
+            // not a sequential read, so stop optimizing for this usage pattern and fall through to the unoptimized behaviour
+            assert streamForSequentialReads.isFullyRead() == false;
+            sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+            IOUtils.close(streamForSequentialReads);
+            streamForSequentialReads = null;
         }
+        return read;
+    }
+
+    /**
+     * If appropriate, open a new stream for sequential reading and satisfy the given read using it.
+     * @return the number of bytes read; if a new stream wasn't opened then nothing was read so the caller should perform the read directly.
+     */
+    private int readFromNewSequentialStream(int part, long pos, byte[] b, int offset, int length)
+        throws IOException {
```

Comment:
The method signature can fit on a single line

Reply:
It didn't always ;) Fixed in 51d2af507dde139718a13d865cb2522b3f0f6a8e